### ISSUE TITLE
It is not possible to place .appex files beforehand in a  particular folder so that they can be placed automatically in Plugins folder of app during the build process(OSX)

### ISSUE THREAD

**Issue by yuktikapahwa** [javafx-maven-plugin/javafx-maven-plugin#235]
There should be a common folder in JavaFX maven plugin for placing the .appex file initially so that it picks it up from there and places it automatically into the Plugins folder during the build process.

Currently , I have to write a script to copy the .appex file inside the Plugins folder after the application is built by the JAVAFX Maven Plugin.
This needs to be resolved.

**Comment by FibreFoX** [comment]
Some reference: http://stackoverflow.com/q/39539031/1961102

Can you describe this a bit more? Where do you want that .appex-files being in what folder, and when? Maybe a small demo-project would be nice to have a more appropriate support from me.

**Comment by FibreFoX** [comment]
Have you looked into the `additionalAppResources`-configuration?
https://github.com/javafx-maven-plugin/javafx-maven-plugin/blob/3d2a4b9e46de4a6e42e925c2de4607e494168994/src/main/java/com/zenjava/javafx/maven/plugin/NativeMojo.java#L210

**Comment by yuktikapahwa** [comment]
Hi ,
I am already using the additionalAppResources tag in the plugin . But using that doesnt solve my problem . The .app which is built using the plugin has a Contents folder which further contains folders like Java ,Plugins,Resources,CodeSingature and MacOS. While the application is built using the javafx maven plugin , it asks to specify the additionalAppResources source and target. The target folder is the target/jfx/app from where the app bundle automatically picks it up and places in the main Java Folder of my app. But I need to place the source (.appex) not inside my Java folder rather inside the Plugins folder of my app (during the app building process not after that) . I am attaching a .png(screenshot) for the Contents folder of my main app bundle (after it gets built from the javafx maven plugin) I am also attaching the sample of my pom where I have specified the java fx maven plugin I am using . Please have a look at it and let me know if there is something wrong in the way i am specifying it.
![screen shot 2016-09-20 at 11 44 18 am](https://cloud.githubusercontent.com/assets/12435224/18677870/4cd5f38c-7f28-11e6-9a75-c2c80b95e38d.png)

[Samplepom.docx](https://github.com/javafx-maven-plugin/javafx-maven-plugin/files/483005/Samplepom.docx)

`

**Comment by FibreFoX** [comment]
Hi, i may have an idea what this is about, but just to be sure: please post the pom-file nox as docx-file-format, the xml-file as is should be okay, or just inline this content. Makes it more secure for me (opening foreign files can be dangerous) and makes it searchable for others having similar questions/problems.

So to sum up, you need to have these appex-files not being inside the application-folder, but as part of you "whole-bundle", right? What Bundler are you using? Do you want to generate some DMG-file or PKG-file (aka "installer")?

**Comment by yuktikapahwa** [comment]
Hi , 
Yes, I cant place this in the application folder because anything which is inside the application folder would straightaway be placed in the Java folder of my main application bundle . But I want it to go it in the Plugins folder of my application . So you need to specify a common location (folder) where I can place all the stuff which can be picked up from there during the build process and straightaway placed in the Plugins folder of my app.
I want to generate .app ,.dmg as well as .pkg . I need all three. 
Just to clarify again , I will explain you what is happening . When the JavaFX maven plugin builds my app as abc.app , I right click on the app to see its Contents (on MAC OSX) . What I see in contents is the following folders -Java ,Plugins,Resources,CodeSingature and MacOS .I have also attached the snapshot for this above. When I go and check in the Plugins folder that is empty . Because the JavaFX Maven Plugin didnt have the workaround to place things in the Plugins folder during the build process. hence it is creating problem for me. I hope I am clear now. I am also attaching the inline xml code for the same.Thanks
[SamplePom.xml.zip](https://github.com/javafx-maven-plugin/javafx-maven-plugin/files/483076/SamplePom.xml.zip)

**Comment by FibreFoX** [comment]
I will take a deep look into placing files outside the "Application" itself on MacOSX-bundlers. If I find something, I will report back. (I think this time I fully understood your idea/feature-request)

**Comment by yuktikapahwa** [comment]
Thankyou so much . I will be waiting to hear back from you soon

**Comment by FibreFoX** [comment]
I might have found a way, bit it gets tricky, and needs proper implementation for being compatible down to 1.8.0_20, don't expect this to be done within the next two days, it is more likely that this will be implemented next week. Or not even be implemented, this is currently just an idea which I have to try first, it even is possible that this will result in some new JDK-bug entry. This feature is not possible out-of-the-box, but might be another point for using this maven-plugin ;)

**Comment by yuktikapahwa** [comment]
Hi ,

Did you get a chance to look into this ?

**Comment by FibreFoX** [comment]
Yes, the solution I was thinking of seems to work, but this week was really hard for me, and I'm implementing some new stuff I'm going to release in the next time.

On the upcoming weekend I hope to find the time to implement this. Sorry for having no feedback for such a long time.

**Comment by FibreFoX** [comment]
If everything works out as planned, this will get available as some snapshot-version tomorrow

**Comment by FibreFoX** [comment]
Hi @yuktikapahwa ,

can you try the current snapshopt-version `8.7.1-SNAPSHOT` ? Please adjust your pom.xml accordingly to the sonatype-repo and further adjustments:
* put your `.appex` files into the folder `src/main/deploy/bundlers/pkg/Contents/PlugIns`
* add to your plugin-configuration `<additionalBundlerResources>src/main/deploy/bundlers</additionalBundlerResources>`
* run `mvn clean package` (when having package-lifecycle-binding)  
or run `mvn clean jfx:native` (when having no lifecycle-binding)

The same goes for the additional `.icns`-files, they should go into `src/main/deploy/bundlers/pkg/Contents/Resources` (or appropiate, not tested ... due to no mac-system available)

**Comment by FibreFoX** [comment]
If this does not work, I might need some other tweaks, just report any problem back here :)

**Comment by FibreFoX** [comment]
Oh, you might need to duplicate some of your files, because every bundler has its own folder (I just re-read this thread to not miss anything)!

**Comment by yuktikapahwa** [comment]
Hi ,

Thankyou for your reply.
I followed exactly the same steps that you mentioned . But it doesnt solve my purpose. Using your snapshot version , what it does is :
It uses the .appex file placed in deploy/bundlers/pkg/Contents/PlugIns from the build directory and places it to the mac temp folder inside a folder called pkg. (in the temp directory ) and doesnt use that at all. The .app file generated by the maven plugin during the build process still doesnt contain the .appex inside it in the PlugIns folder. 
The logs during the build process print the following 

**Found additional bundler resources, trying to copy all files into build root.
[INFO] Copying additional bundler resources into: /var/folders/y3/jv117_75505fnk8htdrs0qm40000gr/T/fxbundler5691789455562613876/images**

But this is not resolving the problem because we want these Plugins and Resources files to be placed inside the app not in any temp location.

**Comment by yuktikapahwa** [comment]
![screen shot 2016-10-05 at 2 12 53 pm](https://cloud.githubusercontent.com/assets/12435224/19125630/d7b1b1a6-8b05-11e6-99f9-61506314e0b5.png)

Please find the screenshot for the folder hierarchy that is created inside the temp location of my mac machine . But the builder doesnt use these files from the temp location to place it inside the .app during the build process. 
Kindly let me know what should be done .

**Comment by FibreFoX** [comment]
Thanks for testing this.

What bundler did you use for this? The "pkg"-bundler? There should have been some output like `"Found additional bundler resources for bundler pkg`.  What happens when you set `<bundler>pkg</bundler>` inside plugin-configuration?

**Comment by yuktikapahwa** [comment]
I am attaching some more screenshots 
1) First is the screenshot of the main application that is built using the javafx maven plugin.
2)Second is the snapshot when we right click on the application bundle and cick on Show Package Contents
3) Third is the Contents folder inside the app
4) Fourth is when we select Contents and check what all it contains .
The PlugIns folder inside the Contents should contain that .appex file and the Resources folder should contain the .icns file . 
But still it is not there.
![screen shot 2016-10-05 at 2 19 52 pm](https://cloud.githubusercontent.com/assets/12435224/19126049/5b99c5ac-8b07-11e6-839f-7e6873abfe7b.png)
![screen shot 2016-10-05 at 2 20 42 pm](https://cloud.githubusercontent.com/assets/12435224/19126052/5ea3c874-8b07-11e6-9be6-07e1ee03431c.png)
![screen shot 2016-10-05 at 2 20 57 pm](https://cloud.githubusercontent.com/assets/12435224/19126055/60d276d6-8b07-11e6-9fad-800cefac6238.png)
![screen shot 2016-10-05 at 2 21 07 pm](https://cloud.githubusercontent.com/assets/12435224/19126057/628d662a-8b07-11e6-832a-92cf66b3bab9.png)

**Comment by FibreFoX** [comment]
Your screenshots are really with it 👍  I didn't realized that mac-applications like this (using the "mac.app"-bundler) behave like "installers". (I really need to find some cheap one ...)

It is not possible to have this working with the "mac.app"-bundler, because it is deleting the target-folder. You are required to use "mac.appStore", "dmg" or "pkg" (these are the internal IDs of the bundlers, so you need to create folders with these names below `src/main/deploy/bundler`).

The ".app"-folder is expected to not work ... sry, but the JDK does not give this feature to us ... I'm glad I found some way for the installers tho.

**Comment by yuktikapahwa** [comment]
So you mean to say , if i add a tag in my plugin configuration as <bundler>pkg</bundler>
or
<bundler>dmg</bundler>

It should work?

**Comment by FibreFoX** [comment]
It only works with installers :/ yes, but you still won't find that files inside that ".app"-folder ...

**Comment by FibreFoX** [comment]
I got some idea about adding some files after the bundler did run ... but it really feels ugly to hack this things ;) but don't worry, I'm used to do such things with the `javapackager`. And having any feedback regarding the installers would be a good sign.

**Comment by yuktikapahwa** [comment]
So I am going to give you a very clear idea of what I did and what output I got. Just to let you know , even specifying the bundler as dmg doesnt resolve my problem . 
I am going to attach a file with the maven build logs and also provide you my plugin confguration .Please have a look and let me know where am I going wrong.

I dont know why I am not allowed to use dmg as a bundler.
And secondly , using pkg as a bundler also doesnt resolve the problem.When I install the app from the pkg bundle , the installed app also doesnt contain the Plugins and Resources files inside it.

So i hope you understand the problem now. Unless these files are not present inside the app , they wont even be a part of the bundler.

I am also attaching a snapshot of my dmg , which will show my app to be dragged inside the applications folder on MAC. When I open the dmg and drag the A.app inside the applications folder and then check that app contents , i dont find the appex files and icns files inside . 

So I hope now you are clear with the problem I am facing . I want these files to be inside my app Contents folder not outside.
![screen shot 2016-10-05 at 3 34 08 pm](https://cloud.githubusercontent.com/assets/12435224/19128408/33f67ee6-8b11-11e6-9deb-2c820d116f21.png)

[logs.docx](https://github.com/javafx-maven-plugin/javafx-maven-plugin/files/512121/logs.docx)
[pom.docx](https://github.com/javafx-maven-plugin/javafx-maven-plugin/files/512122/pom.docx)

**Comment by yuktikapahwa** [comment]
Even if it would have worked with dmg and pkg installers my problem would have been solved .
But the problem here is , even the installers bundle app inside it . So unless the app is complete , the installers will also build with an incomplete app . So our target should be , to get the app built with all the files inside it at desired location.

**Comment by FibreFoX** [comment]
> So i hope you understand the problem now.

I do understand, but just saying that your requested feature isn't even possible with the official tooling provided by the JDK itself. Having anything beyond this takes it time ... and might not be that worth (this does not count for your problem, this really is some nasty feature ....)

>Unless these files are not present inside the app , they wont even be a part of the bundler.

Nope, it should be present, because the installer-bundler works a bit different and does not use the produced ".app"-folder, therefore the bundlers are generating their own ".app"-folder.

It would be nice if you could paste your logs inside this issue with tripple-backticks. Makes it searchable and does not require any software.

**Comment by yuktikapahwa** [comment]
The jar lib/jna-platform-4.2.1.jar has a main class com.sun.jna.Native that does not match the declared main com.aprivacy.application.a
[INFO] Found additional bundler resources, trying to copy all files into build root.
[INFO] Found additional bundler resources for bundler pkg
[INFO] Copying additional bundler resources into: /var/folders/y3/jv117_75505fnk8htdrs0qm40000gr/T/fxbundler6450194527965232568/images/pkg.daemon
![screen shot 2016-10-05 at 4 28 42 pm](https://cloud.githubusercontent.com/assets/12435224/19130351/d19038de-8b18-11e6-9066-d980f3fff652.png)
![screen shot 2016-10-05 at 4 28 27 pm](https://cloud.githubusercontent.com/assets/12435224/19130354/d49b025c-8b18-11e6-805e-fbb62577cf5c.png)


Building PKG package for A



These are the lines which might be useful to you .
While it is building , I go to the respective temp folder location i.e. /var/folders/y3/jv117_75505fnk8htdrs0qm40000gr/T/fxbundler6450194527965232568/images/ on my machine just to check if my pkg is getting built. I could see two folders there , one is the pkg.deamon which had the hierarchy like pkg.deamon/Contents/PlugIns/a.appex and there was another folder called image-4677550100099953218 .This folder had the hierarchy like image-4677550100099953218/a.app

This app was kept at this temp location by the bundler so that it can use this app to build the package.
But the unfortunate part it , the pkg.deamon was outside this app . It was treated as a separate folder .Not as a part of app .I am going to attach a snapshot for the same

**Comment by FibreFoX** [comment]
It sure is possible, that I missed some subfolder ;) thanks for your support and the screenshots, I'm still on analyzing the bundlers, hopefully your input helps me to create another snapshot

**Comment by FibreFoX** [comment]
Could you check what happens when creating "DMG"-installers?

**Comment by yuktikapahwa** [comment]
Could you please confirm what should be added to build a dmg installer . 
I am attaching the pom , I replaced pkg with dmg . Please check if this is the correct way .
Thanks
[pom.docx](https://github.com/javafx-maven-plugin/javafx-maven-plugin/files/512284/pom.docx)

**Comment by FibreFoX** [comment]
Yes, setting `<bundler>dmg</bundler>` should create the application inside some temp-folder, prepare all bundler-files and some .deb-file below `target/jfx/native`.

**Comment by FibreFoX** [comment]
Oh, and make sure the "pkg"-folder now is named "dmg".

**Comment by yuktikapahwa** [comment]
You mean deploy/bundlers/dmg/Contents/PlugIns ?

**Comment by FibreFoX** [comment]
Indeed, renaming that folder from former being "pkg" to "dmg", so the bundler-resource-folder gets detected by the plugin.

**Comment by yuktikapahwa** [comment]
Again the same problem :(
The logs printed the following 
[INFO] Found additional bundler resources, trying to copy all files into build root.
[INFO] Found additional bundler resources for bundler dmg
[INFO] Copying additional bundler resources into: /var/folders/y3/jv117_75505fnk8htdrs0qm40000gr/T/fxbundler5944441384629230023/images/image-3479929629631399350
Building DMG package for A
Preparing Info.plist: /var/folders/y3/jv117_75505fnk8htdrs0qm40000gr/T/fxbundler5944441384629230023/macosx/Info.plist
  Using custom package resource [Bundle config file]  (loaded from package/macosx/Info.plist)
  Using custom package resource [icon]  (loaded from package/macosx/A.icns)

And in the temporary folder , while it is building , I can see the bundle Contents folder outside the app .Please find the snapshot for the following 
![screen shot 2016-10-05 at 5 03 49 pm](https://cloud.githubusercontent.com/assets/12435224/19131605/06519f7c-8b1e-11e6-8527-e9ba7b8e6ca4.png)

**Comment by yuktikapahwa** [comment]
Whereas , the A.app has an internal Contents folder which has PlugIns which should contain A.appex .
But what this maven plugin is doing is : instead of placing it inside , it is creating a Contents folder outside and PlugIns in that which is not our requirement.

**Comment by FibreFoX** [comment]
Thanks for checking, this seems better now, just that I need to get deeper into the "A.app"-folder...but here lies the problem ... the application generated by the pkg/dmg-bundler .... and as one of the first steps, that folder gets deleted, so copying these files beforehand does not work either ...

**Comment by yuktikapahwa** [comment]
That is what I am saying , pasting it beforehand is a problem :(
Because it gets deleted ,it doesnt stay there . 
Moreover I dont want to do any manual pasting stuff because I want the java fx maven plugin to build the complete application bundle for me . :( :( :(

**Comment by FibreFoX** [comment]
Don't worry, I'm eager to find some solution for this :)

Let me check another thing .... found another way for providing some (manipulated) application, but don't let your hopes be too high ... this area of the bundlers just does not exist and I'm working within thin air.

**Comment by yuktikapahwa** [comment]
Sure .
Thankyou so much for all the help .

**Comment by FibreFoX** [comment]
As this is a spare-time project, and because I need some sleep, I will report back to you after having this reworked a bit more. Thanks for your patience and the support for getting this feature working.

**Comment by yuktikapahwa** [comment]
Sure. Not a problem .
I will wait to hear back from you.
Thanks again

**Comment by FibreFoX** [comment]
Hi @yuktikapahwa ,

I have uploaded another snapshot-version, which is highly experimental and sourcecode is not yet published to github yet. Can you recheck again?

You will need to change the foldername to "mac.app" this time ...

**Comment by yuktikapahwa** [comment]
Shall i use the same snapshot version i.e. 8.7.1-SNAPSHOT?
Also which folder name should be mac.app?
What I understand is src/main/deploy/bundlers/mac.app/Contents/PlugIns . Is that correct?

**Comment by yuktikapahwa** [comment]
It should be mac.app or my application name.app ?

**Comment by FibreFoX** [comment]
Yes, same version and "mac.app", because thats the name of the bundler.

**Comment by yuktikapahwa** [comment]
Also , in the pom , should i use - 
<bundler>mac.app</bundler>

**Comment by yuktikapahwa** [comment]
Please mention the exact bundler tag to be used in the pom

**Comment by FibreFoX** [comment]
Good question ;) i forgot about that detail. Please remove that line, all bundlers should work. I am using `<additionalBundlerResources>` for the provisioning of the generated "mac.app" too in addition to `<additionalAppResources>`.

**Comment by yuktikapahwa** [comment]
So , in this case do you expect the bundler resources to be copied in all the installers in case of Mac i.e. app , dmg as well as pkg ?

**Comment by FibreFoX** [comment]
Thats right, because I manipulate the resources. Within the mac-bundlers, there is a secret switch to point to some "pre defined" (or better already build) application. Therefore when running without any single bundler selected, the "mac.app"-bundler runs first, creating all the files inside the native-folder, now being enhanced with the `additionalBundlerResources` too. The next mac-bundler which gets called, should pick up the generated application-bundle at `target/jfx/native/{appName}.app`. This way the application-bundle created by the installers won't re-create the bundle but instead use the already existing/produced one.

**Comment by yuktikapahwa** [comment]
Please give me some time . There is some wierd thing I can see in my logs using this version of maven plugin due to which the build breaks .
Earlier , this plugin was first building the app and then  codesigning everything inside the app with developer id application certificate, then again generating the app to the temp folder and then codesigining the app with developer id application certificate and then generating a dmg out of it in the target/jfx/native location , and then again generating an app in temp location , codesigning evertything inside the app with Developer id application certificate and building a pkg out of it .

But now what is happening is , it build app in target/jfx/native , then codesigns everything inside the app with Developer id application cvertificate , then use the same app to build the dmg and place that inside the target/jfx/native location and then it tries to use the same app to generate the pkg out of it , but this time it is seaching for Developer Id Installer certificate keys to build the package which it was not doing before . 

So did you change anything in the way the pkg bundler is built using the plugin . Did you change the certificate from Developer Id application to Developer Id installer .

Because it is not able to find the keys for installer certificate , it prints the foloowing
Config files are saved to /var/folders/y3/jv117_75505fnk8htdrs0qm40000gr/T/fxbundler4914586228054903100/macosx. Use them to customize package.
[INFO] Found additional bundler resources, trying to copy all files into build root.
[INFO] Copying additional bundler resources into: /var/folders/y3/jv117_75505fnk8htdrs0qm40000gr/T/fxbundler5721012892434601018/images
Building PKG package for A
  Using default package resource [pkg background image]  (add package/macosx/A-background.png to the class path to customize)
Preparing distribution.dist: /var/folders/y3/jv117_75505fnk8htdrs0qm40000gr/T/fxbundler5721012892434601018/macosx/distribution.dist
Using default package resource [script to run after application image is populated]  (add package/macosx/A-post-image.sh to the class path to customize)
Running [pkgbuild, --component, /Users/y/Documents/workspace_Feb1/osxCoreInstaller/target/jfx/native/A.app, --install-location, /Applications, /var/folders/y3/jv117_75505fnk8htdrs0qm40000gr/T/fxbundler5721012892434601018/packages/A-app.pkg]
pkgbuild: Adding component at /Users/y/Documents/workspace_Feb1/osxCoreInstaller/target/jfx/native/A.app
pkgbuild: Wrote package to /var/folders/y3/jv117_75505fnk8htdrs0qm40000gr/T/fxbundler5721012892434601018/packages/A-app.pkg
Running [security, find-certificate, -c, Developer ID Installer: , -a]
Did not find a key matching 'Developer ID Installer: '
Running [productbuild, --resources, /var/folders/y3/jv117_75505fnk8htdrs0qm40000gr/T/fxbundler5721012892434601018/macosx, --distribution, /var/folders/y3/jv117_75505fnk8htdrs0qm40000gr/T/fxbundler5721012892434601018/macosx/distribution.dist, --package-path, /var/folders/y3/jv117_75505fnk8htdrs0qm40000gr/T/fxbundler5721012892434601018/packages, /Users/yuktikapahwa/Documents/workspace_Feb1/osxCoreInstaller/target/jfx/native/A-1.0.pkg]
productbuild: Wrote product to /Users/yuktikapahwa/Documents/workspace_Feb1/osxCoreInstaller/target/jfx/native/A-1.0.pkg
  Config files are saved to /var/folders/y3/jv117_75505fnk8htdrs0qm40000gr/T/fxbundler5721012892434601018/macosx. Use them to customize package.
[INFO] Skipping Mac App Store Ready Bundler because of configuration error Specified image directory mac.app.image: /Users/yuktikapahwa/Documents/workspace_Feb1/osxCoreInstaller/target/jfx/native/A.app does not exists
Advice to Fix: Confirm that the value for mac.app.image exists
[INFO] Skipping WebStart JNLP Bundler because of configuration error No OutFile Specificed
Advice to Fix: Please specify the name of the JNLP Outut file in 'jnlp.outfile'



As you can see in the logs , it prints skipping mac app store ready bundler .Previously it was taking the 3rd party developer certificate keys from my key chain and was also generating the mac app store bundler package . But this time it excluded that also .

**Comment by FibreFoX** [comment]
I do have changed something, but nothing signing related ... at least I thought. It is possible that signing gets skipped when having "prepared application", which makes sense to not re-sign them over and over again.

Can you check that you have set `src/main/deploy/bundlers/mac.app/` and not `src/main/deploy/bundlers/A.app/` ? I'm missing some output.

But it still is not right :) will work on it a bit.

btw: to have good code-escaping on github, just use triple-backticks ``` before and after the block you want to have special text-layout

**Comment by yuktikapahwa** [comment]
Now I am going two explain you two very major issues with this change you have made .I can tell this to you with my experience 
1) What I can see during this build process , that firstly the .app gets generated , then everything inside it gets codesigned and then it copies the additional bundle resources inside the app and then using the new app it generates the dmg and pkg.
But this is not how Mac works . If I use this app and upload it to my website and then any user download this from our website , Mac will not allow the user to use this app.It will throw an error saying that the app is corrupted kindly move it to trash .
So for you to understand , It means that once the app is built using the plugin and codesigned you cannot copy anything inside it , which you are doing , because of which the app gets corrupted. I am sorry to say but this is how Mac works .It is really strict about codesign and we cannot tamper with it .
So , if you are making some change in the plugin to set the bundler configuration , it should be in a way that you should copy everything inside the app first and then codesign it rather than first codesigning it and then copying the bundler resources.
2) Secondly , I have already mentioned this issue previously , that with your new build  somehow the  plugin is not able to sign the app witH 3rd party distribution certificate to distribute it to the app sore . It prints something like this in the logs 
```[INFO] Skipping Mac App Store Ready Bundler because of configuration error Specified image directory mac.app.image: /Users/yuktikapahwa/Documents/workspace_Feb1/osxCoreInstaller/target/jfx/native/APrivacy.app does not exists
Advice to Fix: Confirm that the value for mac.app.image exists```

With the previous version of javafx maven plugin , this was not the case .Previous version used to prepare the app then codesign it with Developer ID certificate and then build the dmg and pkg and after that again build a MacApp store pkg with 3rd party certificate.
Now even that part has broken .

Please let me know in case any other information is required

**Comment by FibreFoX** [comment]
Thats sad to hear ... yes, I mostly thought about only the native files are getting signed ... this is clearly some lack of experience in developing native mac-applications!

To make it clear: I'm doing magic here! You are requiring something that does not exist ... and everything I'm now experiencing while developing are throwbacks ... because this just isn't possible when using current existing bundlers.

This gets me to the point where anything using the JDK mac-bundlers results in some dead end. I really have to think about this .... there IS some option, but I'm not sure about the license-situation (as I am required to copy most of the code from the JDK) and how to provide your requested feature. I got some feedback from other developers using this feature, but they are working on windows-systems.

Yes, it is depressing ... and we came that far .... I will have to go back and reconsider the whole process.

**Comment by yuktikapahwa** [comment]
I know :(
You tried so hard but trust me , when you are working with Mac , you reach such dead end .This is nothing new for me :(

**Comment by FibreFoX** [comment]
The solution I'm currently thinking of: extending the mac-bundler classes and overriding the critical paths with the additional changes required for having additional bundler files. I'm unsure if this makes the created files being to released under GPLv2 or not ...

The solution might work, because I've added the ability to add custom self-made bundlers some releases ago. But it might require you to skip all default-bundlers from the JDK, which currently isn't existing inside the javafx-maven-plugin. I will need to add some "onlyCustomBundlers"-parameter to make them being skipped/ignored. All Mac-bundlers are suspect of being re-written, mostly because the installer-bundlers are depending on the application-bundler (where all the stuff has to be signed in the first place).

**Comment by FibreFoX** [comment]
Hi @yuktikapahwa , after some sleep and bad dreams about it, I came to the final conclusion, that I will remove all the workarounds developed for this issue!

Don't panic ;) this does NOT mean that you are stranded here, because the solution about re-using the already existing bundlers is my prefered way. This will take some time, because I'm actually working on two plugins in parallel: [javafx-gradle-plugin](https://github.com/FibreFoX/javafx-gradle-plugin) and this javafx-maven-plugin.

Just to make sure, I want to share my ideas about it: when I understood you correct, it should work when all additional files are copied into that ".app"-folder BEFORE any signing happens, right? So all I should have to do is to add file-copy-action before `MacBaseInstallerBundler.signAppBundle` is called, which happens after the file-generations, right?

**Comment by FibreFoX** [comment]
This issue will definitely result in some JDK-bugtracker entry, just let me finish this for being able to show the difference inside that bug-report. It really is very interesting to learn all this issues and workarounds, makes theses plugins better with every issue-report! Even if it is as depressing like this one, I won't give up that easily, went through worse while developing that nasty workaround for the javafx-gradle-plugin where I needed to rewrite on bytecode-level because some leaked file-handler.

Will report back with a new snapshot-version ;)

**Comment by yuktikapahwa** [comment]
Hi @FibreFoX  Yes your understanding is absolutely correct . We need to call the ```MacBaseInstallerBundler.signAppBundle``` after doing all the copy pasting stuff inside the application. 
Also make sure , while trying to find a workaround for this , the Mac App Store Ready Bundler should not be affected. For Mac App store ready bundler also you should make sure that the pkg generated out of it should contain the app with all the latest PlugIns and Resources. 
So whether the ```MacBaseInstallerBundler.signAppBundle``` should sign the bundle with Developer ID or 3rd party certificate , it should sign the complete bundle not the one without the Mac App Bundlers.

Thank you so much for your time. Please let me know in case you require any further detailed information .

**Comment by FibreFoX** [comment]
The `MacAppStoreBundler` WILL GET AFFECTED because I'm exchanging the used application-bundler. The problem with signing might be something I messed up with my last changes. All I'm adding is the additional step of copying these files into the folders BEFORE signing-process happens, regardless of 3rd-party or not.

The bundler HAS to be extended of the existing `MacAppBundler`-class, and will be present in the next snapshot-version. There is nothing that has to be added to the dependencies, this is some core-feature only not-working for mac, so I decided to add this extending class to the core of javafx-maven-plugin. The licence-thing is no issue because of the classpath-exclusion of the GPL stated in these files. It only gets complicated with the whole loop over all bundlers and the detection about the need for parameter-manipulations. What bugs me more is the version-compatibility, I need to make sure this even works with 1.8.0_20 ;) because thats the lowest version this plugin requires.

Don't worry, I am on this and you don't need to implement your own bundler for this (because I am doing this right now).

**Comment by yuktikapahwa** [comment]
Thank you so much :)
Take your time !

**Comment by FibreFoX** [comment]
Changes are implemented on the javafx-gradle-plugin (https://github.com/FibreFoX/javafx-gradle-plugin/commit/65cef6dd24ab6207b6f1aec8866a39ec33daf0e9), I'm going to migrate the changes to the javafx-maven-plugin and then create some snapshot-version, but the MAIN work is done now ;) at least I hope.

**Comment by FibreFoX** [comment]
@yuktikapahwa I finally migrated the changes, can you recheck with the current snapshot? It's again the same version `8.7.1-SNAPSHOT`, hopefully this works out for you now. At least on linux/windows I seem to not have broke something.

**Comment by yuktikapahwa** [comment]
Hi @FibreFoX  It is a long weekend here . I will be able to join once I am back to work on Tuesday. I will get back to you once I am done testing.

**Comment by FibreFoX** [comment]
Hi @yuktikapahwa , did you get some time to check?

**Comment by yuktikapahwa** [comment]
Hi @FibreFoX  ...So please tell me the changes i need to make in my pom .Do i need to use bundlers and what folders should i create inside the deploy folder.

**Comment by FibreFoX** [comment]
- Adjustments to your `pom.xml`
1. adding sonatype-repo
2. using `8.7.1-SNAPSHOT` as version of the javafx-maven-plugin
3. setting `additionalBundlerResources` to your special folder 

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <packaging>jar</packaging>

<!-- ... -->
<pluginRepositories>
    <pluginRepository>
        <id>oss-sonatype-snapshots</id>
        <url>https://oss.sonatype.org/content/groups/public/</url>
        <snapshots>
            <enabled>true</enabled>
        </snapshots>
    </pluginRepository>
</pluginRepositories>
<!-- ... -->

    <build>
        <plugins>
            <plugin>
                <groupId>com.zenjava</groupId>
                <artifactId>javafx-maven-plugin</artifactId>
                <version>8.7.1-SNAPSHOT</version>
                <configuration>
                    <verbose>true</verbose>
                    <!-- ... -->
                    <bundleArguments>
                        <mac.app.additionalBundlerResources>src/main/deploy/bundlers/mac.app</mac.app.additionalBundlerResources>
                    </bundleArguments>

                    <additionalBundlerResources>src/main/deploy/bundlers</additionalBundlerResources>
                </configuration>
                <executions>
                    <execution>
                        <!-- required before build-native -->
                        <id>create-jfxjar</id>
                        <phase>package</phase>
                        <goals>
                            <goal>build-jar</goal>
                        </goals>
                    </execution>
                    <execution>
                        <id>create-native</id>
                        <phase>package</phase>
                        <goals>
                            <goal>build-native</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
        </plugins>
    </build>
</project>
```
**Please note:** you don't have to specify any bundler here, but please keep all your additional information about your signing-keys and other stuff

# Special folder locations and their content
1. create `src/main/deploy/bundlers/`-folder
2. when running und Mac, you are required to create a folder: `src/main/deploy/bundlers/mac.app/`
3. place your additional files into that new folder, you will have something like this now: `src/main/deploy/bundlers/mac.app/Contents/PlugIns`

Now you can create the native app and all installers (aside of the JNLP-files) with calling `mvn clean package` (I always do CLEAN my workspace to avoid faulty data)

**Comment by FibreFoX** [comment]
@yuktikapahwa any progress on this?

If you know any CI-server which is for free for open-source projects, it wouldn't need to nag you about ;) sry

**Comment by yuktikapahwa** [comment]
Hi @FibreFoX  I am really sorry for not getting back to you . I am stuck at something really important at work which is a high priority .
Please give me a day more and I will test it and surely message you back .
Thank you for your patience.

**Comment by FibreFoX** [comment]
Don't worry, I'm just eager to finally have this nailed down ;) take your time!

**Comment by FibreFoX** [comment]
I just realized, that travis-ci is offering OSX as built-system too .... damn, thats what I need!
https://docs.travis-ci.com/user/multi-os/

**Comment by FibreFoX** [comment]
Hi @yuktikapahwa, did you get the chance to check the latest SNAPSHOT? Or do you know when you get the time for it?

**Comment by yuktikapahwa** [comment]
Hi @FibreFoX  I am really really sorry .These days I am stuck on to one big issue at work and I am not finiding time to do anything else. I am waiting for this issue to get sorted out so that I can proceed with remaining things . I am pretty sure I will be able to test it by friday ! 100 %

**Comment by FibreFoX** [comment]
Okay :) thanks for the update. I will wait until next monday for some feedback, maybe I can find something to check it myself (but chances are not good).

**Comment by FibreFoX** [comment]
@yuktikapahwa Any progress on this one? Or do you know how I can check this for myself, is there any file being generated where I can check the additional files inside the plugin-folder are getting picked up while signing? (As some reminder: I totally do not know how mac-development works)

**Comment by yuktikapahwa** [comment]
Could you please tell me why it still couldn't copy the bundler resources .It prints something like this in the logs during building 
`[INFO] Found additional bundler resources, trying to copy all files into build root, using:/Users/yuktikapahwa/Documents/workspace_Feb1/osxCoreInstaller/src/main/deploy/bundlers
[INFO] No additional bundler resources for bundler mac.app was found, copying all files instead.
[INFO] Copying additional bundler resources into: /var/folders/y3/jv117_75505fnk8htdrs0qm40000gr/T/fxbundler6689367430911457610/images
[WARNING] Couldn't copy resource /Users/yuktikapahwa/Documents/workspace_Feb1/osxCoreInstaller/src/main/deploy/bundlers with reason /Users/yuktikapahwa/Documents/workspace_Feb1/osxCoreInstaller/src/main/deploy/bundlers`

**Comment by yuktikapahwa** [comment]
What is the expectation this time ?
Like while it starts building the plugin generates the app first then dmg and finally pkg. So do you expect the plugins to go inside the app before hand with the changes you have made?

**Comment by yuktikapahwa** [comment]
Below is my pom
`	<plugin>
				<groupId>com.zenjava</groupId>
				<artifactId>javafx-maven-plugin</artifactId>
				<version>8.7.1-SNAPSHOT</version>
				<configuration>
					<mainClass>com.a.application.A</mainClass>
					<deployDir>${project.basedir}/deploy</deployDir>
					<needShortcut>true</needShortcut>
					<appName>A</appName>
					<jfxMainAppJarName>a-jfx.jar</jfxMainAppJarName>

					
					<verbose>true</verbose>
 <bundleArguments>
                        <mac.app.additionalBundlerResources>src/main/deploy/bundlers/mac.app</mac.app.additionalBundlerResources>
                    </bundleArguments>
					
					<jfxAppOutputDir>${project.build.directory}/jfx/app</jfxAppOutputDir>

					
					<additionalAppResources>${project.build.directory}/resources/macosx/</additionalAppResources>
					
                    <additionalBundlerResources>src/main/deploy/bundlers</additionalBundlerResources>
				</configuration>

				
				<executions>

					

					<execution>
						<id>create-jfxjar</id>
						<phase>package</phase>
						<goals>
							<goal>build-jar</goal>
						</goals>
					</execution>

					
					 <execution>
						<id>create-native</id>
						<phase>package</phase>
						<goals>
							<goal>build-native</goal>
						</goals>
					</execution>  
				</executions>
			</plugin>`

**Comment by yuktikapahwa** [comment]
`<?xml version="1.0" encoding="UTF-8"?>
<plugin>
   <groupId>com.zenjava</groupId>
   <artifactId>javafx-maven-plugin</artifactId>
   <version>8.7.1-SNAPSHOT</version>
   <configuration>
      <mainClass>com.a.application.A</mainClass>
      <deployDir>${project.basedir}/deploy</deployDir>
      <needShortcut>true</needShortcut>
      <appName>A</appName>
      <jfxMainAppJarName>a-jfx.jar</jfxMainAppJarName>
      
      <verbose>true</verbose>
      <vendor>${target.publisher}</vendor>
      <bundleArguments>
         <mac.app.additionalBundlerResources>src/main/deploy/bundlers/mac.app</mac.app.additionalBundlerResources>
      </bundleArguments>
      
      <jfxAppOutputDir>${project.build.directory}/jfx/app</jfxAppOutputDir>
      
      <additionalAppResources>${project.build.directory}/resources/macosx/</additionalAppResources>
      <additionalBundlerResources>src/main/deploy/bundlers</additionalBundlerResources>
   </configuration>
   
   <executions>
      
      <execution>
         <id>create-jfxjar</id>
         <phase>package</phase>
         <goals>
            <goal>build-jar</goal>
         </goals>
      </execution>
      
      <execution>
         <id>create-native</id>
         <phase>package</phase>
         <goals>
            <goal>build-native</goal>
         </goals>
      </execution>
   </executions>
</plugin>'

**Comment by yuktikapahwa** [comment]
I am attaching the pom settings for java fx maven plugin .Kindly look into it and let me know if anything is missed
[pom.docx](https://github.com/javafx-maven-plugin/javafx-maven-plugin/files/548090/pom.docx)

**Comment by FibreFoX** [comment]
Thanks for looking into this again :)

What folder-stucture do you have below src/main/deploy/bundlers? It is required to have some "mac.app"-folder where your additional files are in.

**Comment by yuktikapahwa** [comment]
![screen shot 2016-10-24 at 11 34 52 am](https://cloud.githubusercontent.com/assets/12435224/19652218/037f04d2-99de-11e6-82ee-56ed5630e70d.png)

**Comment by yuktikapahwa** [comment]
hI @FibreFoX 
Were you able to figure out something?

**Comment by FibreFoX** [comment]
Nope not yet, I'm currently trying to set up some virtual machine :( 

Your settings are correct and your folder-path looks okay ... just caching some old plugin-version might be possible, but I doubt this :/ so I need to verify this myself ... maybe something stupid I missed again. Will report back later ...

**Comment by FibreFoX** [comment]
Hi @yuktikapahwa , I have found some mistake :) can you re-try now? Please don't worry, this snapshot-build is a bit verbose, will be reduced after you can verify this now working.

**Comment by yuktikapahwa** [comment]
Sure, let me try and get back to you.

**Comment by yuktikapahwa** [comment]
`[INFO] Found additional bundler resources, trying to copy all files into build root, using:/Users/yuktikapahwa/Documents/workspace_Feb1/osxCoreInstaller/src/main/deploy/bundlers
[INFO] No additional bundler resources for bundler mac.app was found, copying all files instead.
[INFO] Copying additional bundler resources into: /var/folders/y3/jv117_75505fnk8htdrs0qm40000gr/T/fxbundler8322738560364683438/images
[WARNING] Couldn't copy resource /Users/yuktikapahwa/Documents/workspace_Feb1/osxCoreInstaller/src/main/deploy/bundlers with reason /Users/yuktikapahwa/Documents/workspace_Feb1/osxCoreInstaller/src/main/deploy/bundlers
Using special javafx-maven-plugin MacAppBundler
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 21.938 s
[INFO] Finished at: 2016-10-25T09:07:34-04:00
[INFO] Final Memory: 29M/405M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.zenjava:javafx-maven-plugin:8.7.1-SNAPSHOT:build-native (create-native) on project osxCoreInstaller: An error occurred while generating native deployment bundles: Got exception while copying additional bundler resources: /Users/yuktikapahwa/Documents/workspace_Feb1/osxCoreInstaller/src/main/deploy/bundlers -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.'

**Comment by yuktikapahwa** [comment]
`[INFO] Found additional bundler resources, trying to copy all files into build root, using:/Users/yuktikapahwa/Documents/workspace_Feb1/osxCoreInstaller/src/main/deploy/bundlers
[INFO] No additional bundler resources for bundler mac.app was found, copying all files instead.
[INFO] Copying additional bundler resources into: /var/folders/y3/jv117_75505fnk8htdrs0qm40000gr/T/fxbundler8322738560364683438/images
[WARNING] Couldn't copy resource /Users/yuktikapahwa/Documents/workspace_Feb1/osxCoreInstaller/src/main/deploy/bundlers with reason /Users/yuktikapahwa/Documents/workspace_Feb1/osxCoreInstaller/src/main/deploy/bundlers
Using special javafx-maven-plugin MacAppBundler`

**Comment by yuktikapahwa** [comment]
`[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 21.938 s
[INFO] Finished at: 2016-10-25T09:07:34-04:00
[INFO] Final Memory: 29M/405M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.zenjava:javafx-maven-plugin:8.7.1-SNAPSHOT:build-native (create-native) on project osxCoreInstaller: An error occurred while generating native deployment bundles: Got exception while copying additional bundler resources: /Users/yuktikapahwa/Documents/workspace_Feb1/osxCoreInstaller/src/main/deploy/bundlers -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.`

**Comment by FibreFoX** [comment]
Do you have verbose-flag being enabled?

**Comment by FibreFoX** [comment]
I tweaked Travis-CI a bit, you should get something like this:
https://travis-ci.org/javafx-maven-plugin/javafx-maven-plugin/jobs/170421331#L3228

As my system is not capable of running macos inside virtualbox, I'm lacking proper development-systems :( sorry that it takes that many attempts

**Comment by yuktikapahwa** [comment]
Check the difference for what is printed for you and what is printed in my logs. 
These are your logs :
` [INFO] Found additional bundler resources for bundler mac.app
[INFO] [INFO] Skipped copying additional bundler resources, mostly because this bundler does not need them. You might want to use additionalAppResources. To make sure, check for any warnings printed above this message.
[INFO] Using special javafx-maven-plugin MacAppBundler`

These are my logs 
`[INFO] Found additional bundler resources, trying to copy all files into build root, using:/Users/yuktikapahwa/Documents/workspace_Feb1/osxCoreInstaller/src/main/deploy/bundlers
[INFO] No additional bundler resources for bundler mac.app was found, copying all files instead.
[INFO] Copying additional bundler resources into: /var/folders/y3/jv117_75505fnk8htdrs0qm40000gr/T/fxbundler8750730379711139870/images
[WARNING] Couldn't copy resource /Users/yuktikapahwa/Documents/workspace_Feb1/osxCoreInstaller/src/main/deploy/bundlers with reason /Users/yuktikapahwa/Documents/workspace_Feb1/osxCoreInstaller/src/main/deploy/bundlers
Using special javafx-maven-plugin MacAppBundler`

**Comment by yuktikapahwa** [comment]
It is trying to copy something in my temp directory and prints telling that it couldn't copy the resource. This doesnt happen for you though !

**Comment by FibreFoX** [comment]
Will try to cleanup the messages and will report back with new version. Just in theory, when the signing process applies, is there a way to check the application bundle via mac-commands? Maybe I can check this via test-projects running on travis. I just need something to verify this is working, now that I can run travis with mac os.

**Comment by yuktikapahwa** [comment]
Using mac commands? Meaning shell scripts?

**Comment by FibreFoX** [comment]
Yes, I need to make sure that this feature is verifiable via testing projects. As this isn't something `jarsigner` related, and lacking some Mac, I need something that can be executed to verify the signed application, this will be done on travis.

**Comment by FibreFoX** [comment]
Typical case of "RTFM" :D
http://www.manpagez.com/man/1/codesign/
`codesign -v [-R requirement] [-v] [path|pid ...]`

Will try to add this into the IT project.

**Comment by yuktikapahwa** [comment]
I used this command to check the signage of the application 
codesign -dvvv <path to the application>

**Comment by yuktikapahwa** [comment]
Hi @FibreFoX 
Any progress on this one ?

**Comment by FibreFoX** [comment]
Nope, and pretty impossible, because I would require to have some secret-value, but this does not work for pull-requests, which means this idea of having travis as support for this simply does not work. I'm currently on moving some parts to the Workarounds-class, some cleanup and removing verbosity when not required. Maybe I will find some mind-quirks while doing this. I will push another SNAPSHOT-version when I'm finished with that.

**Comment by FibreFoX** [comment]
One thing: ` [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch`
Could you re-run your project (by providing `-e` parameter) and copy-paste the stacktrace? Maybe there was some problem with the way I'm calling some private methods :( which I had to call via reflection (that bundler isn't a nice thing for inheritance)

**Comment by yuktikapahwa** [comment]
Can u please specify in detail ....where should i add the -e parameter ? Where exactly in my pom file

**Comment by FibreFoX** [comment]
It is while calling maven ;) not a special configuration. It makes maven do verbose output of the stacktrace, which might help me to find this problem (probably something with the way I'm using reflection)

**Comment by FibreFoX** [comment]
Hi @yuktikapahwa , I finally got my own MacOSX-setup, so this issue will get being worked on again ;)

**Comment by yuktikapahwa** [comment]
Hi @FibreFoX  , Thanks for getting back to me .
Could you please let me know how can i test the same and where lies the updated version of this javafx maven plugin .
Kindly let me know the steps in detail.
Thankyou

**Comment by FibreFoX** [comment]
You might have misunderstood me ;) I am working on it now (because I have finally some setup where I can run some MacOSX), when I have something for you to verify, there will be instructions here on this issue. Might take a while, but I am working on it now with proper verification methods

**Comment by FibreFoX** [comment]
Hi @yuktikapahwa ,

I finally made some progress here, would be nice if YOU can check this too.

# Adjustments to your `pom.xml`
1. adding sonatype-repo
2. using `8.8.0-SNAPSHOT` as version of the javafx-maven-plugin
3. setting `additionalBundlerResources` to your special folder 

``` xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <packaging>jar</packaging>

<!-- ... -->
<pluginRepositories>
    <pluginRepository>
        <id>oss-sonatype-snapshots</id>
        <url>https://oss.sonatype.org/content/groups/public/</url>
        <snapshots>
            <enabled>true</enabled>
        </snapshots>
    </pluginRepository>
</pluginRepositories>
<!-- ... -->

    <build>
        <plugins>
            <plugin>
                <groupId>com.zenjava</groupId>
                <artifactId>javafx-maven-plugin</artifactId>
                <version>8.8.0-SNAPSHOT</version>
                <configuration>
                    <verbose>true</verbose>
                    <!-- ... -->
                    <additionalBundlerResources>src/main/deploy/bundlers</additionalBundlerResources>
                </configuration>
                <executions>
                    <execution>
                        <!-- required before build-native -->
                        <id>create-jfxjar</id>
                        <phase>package</phase>
                        <goals>
                            <goal>build-jar</goal>
                        </goals>
                    </execution>
                    <execution>
                        <id>create-native</id>
                        <phase>package</phase>
                        <goals>
                            <goal>build-native</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
        </plugins>
    </build>
</project>
```

**Please note:** you don't have to specify any bundler here, but please keep all your additional information about your signing-keys and other stuff

# Special folder locations and their content
1. create `src/main/deploy/bundlers/`-folder
2. when running on Mac, you are required to create a folder: `src/main/deploy/bundlers/mac.app/`
3. place your additional files into that new folder, you will have something like this now: `src/main/deploy/bundlers/mac.app/Contents/PlugIns`

Now you can create the native app and all installers (aside of the JNLP-files) with calling `mvn clean package`

Only the signing-part wasn't checked by me, so I require your help here. If this finally works, I will push my changes here along with some other things, following by some release of the new version.

**Comment by yuktikapahwa** [comment]
Hey  @FibreFoX. ..thankyou for your effort on this issue. I can not try it before Monday. 

Once I try it I will for sure get back to you with results. 
Thankyou

**Comment by FibreFoX** [comment]
Hi @yuktikapahwa , any luck to get this tested with your app?

**Comment by FibreFoX** [comment]
ping @yuktikapahwa

**Comment by FibreFoX** [comment]
Hey @yuktikapahwa, I have merged the changes, as there is not further feedback, I suspect you are still busy with other things. As this is working in my MacOSX setup, I suspect this working. Please feel free to provide further feedback on this, will hold this issue open for two more weeks.

**Comment by FibreFoX** [comment]
Hey @yuktikapahwa, I have merged the changes, as there is not further feedback, I suspect you are still busy with other things. As this is working in my MacOSX setup, I suspect this working. Please feel free to provide further feedback on this, will hold this issue open for two more weeks.

**Comment by yuktikapahwa** [comment]
Hi @FibreFoX  , I am really sorry for not getting back to you. You suspected it correct. I was really busy with other things and I am done with those now.
I am going to try your changes right away.
Let me do that and hopefully give you a positive news.
Thankyou .

**Comment by FibreFoX** [comment]
No problem :D I got some patience while developing/maintaining this maven-plugin. Hopefully this now really does work!

**Comment by yuktikapahwa** [comment]
So , I tried out everything that is mentioned.
I have one good news and one bad. The good is that , whatever I place inside the bundlers/mac.app/Contents/ folder it gets automatically picked up during the app creation and gets placed there.So I placed basically two things there : one is PlugIns/A.appex and another is Resources/a.icns

While the maven plugin was building the app , it placed everything at the correct position , and then starts code signing every single component inside it . It goes through every folder inside the app i.e. Java ,MacOSX,PlugIns ,Resources. Everything gets codesigned properly. When it reaches PlugIns folder , it says the .appex file is already signed hence cant codesign it . Whereas I make sure ,that the .appex is not codesigned at all previously by any certificate. 
Because of this , the codesign by the javafx maven plugin breaks and the main app fails to codesign . Similary it happens while building the dmg and the package. None of them gets built by the maven plugin because the codesign fails.
I am showing the logs below, if by looking at them you can find out what is going around and why did the javafx maven plugin failed to sign the .appex file.

`java.io.IOException: Exec failed with code 1 command [[codesign, -s,  Application: A LTD. (YYYYYYYYYY), --prefix, com.a.z., -vvvv, /Users/a/Documents/workspace/osxInstaller/target/jfx/native/A.app/Contents/PlugIns/AMenu.appex] in unspecified directory
	at com.oracle.tools.packager.IOUtils.exec(IOUtils.java:165)
	at com.oracle.tools.packager.IOUtils.exec(IOUtils.java:138)
	at com.oracle.tools.packager.IOUtils.exec(IOUtils.java:132)
	at com.oracle.tools.packager.mac.MacBaseInstallerBundler.lambda$signAppBundle$191(MacBaseInstallerBundler.java:307)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:580)
	at com.oracle.tools.packager.mac.MacBaseInstallerBundler.signAppBundle(MacBaseInstallerBundler.java:329)
	at com.oracle.tools.packager.mac.MacBaseInstallerBundler.signAppBundle(MacBaseInstallerBundler.java:219)
	at com.zenjava.javafx.maven.plugin.MacAppBundlerWithAdditionalResources.doBundle(MacAppBundlerWithAdditionalResources.java:219)
	at com.oracle.tools.packager.mac.MacAppBundler.execute(MacAppBundler.java:1096)
	at com.zenjava.javafx.maven.plugin.NativeMojo.execute(NativeMojo.java:647)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:863)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:199)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
`

Do u have any idea why this is happening . The whole objective was to place the appex file in the plugins folder before hand and then successfully build the app. But now although it gets placed beforehand , it still fails to build a properly codesigned application , which is of no use. The moment I will take this failed code signed app to some other machine , it will be found as corrupted becuase it is not codesigned.

**Comment by FibreFoX** [comment]
The thing is, that I wrote my very own bundler, which behaves mostly like the one from the JDK, but additionally copied that new files ... makes me very curious about that signing-part. The new bundler replaces the default one, therefor it is expected that the installers are failing, because they are themselves doing one application build.
But I am happy, because we finally are getting close! I will try to get my mac-setup to do code-signing too ... somehow. If you by chance find it first, I would be happy to read about.

**Comment by FibreFoX** [comment]
By any chance: do you have some expired certificate for that ID?
http://ericholsinger.com/programming/xcode-codesign-failed-with-exit-code-1-expired-certificates-causing-ambiguous-matches/

;) or did you try restarting? http://stackoverflow.com/questions/19585253/codesign-error-with-xcode5-0-1-mac-os-10-9

**Comment by FibreFoX** [comment]
@yuktikapahwa To make sure it would work "normally", you can try to create the application, and manually call codesign afterwards. Maybe you will get more output and see something which we are missing here.

**Comment by yuktikapahwa** [comment]
My certificates are expiring in the year 2021 and restarting also didnt work.
Manual is not an option becuase when the javafx maven plugin is building it , it takes the app and codesigns it then it builds the dmg.But once if the codesigning of the app breaks , the plugin fails to build anything .Neither the dmg not the pkg.
I have no workaround for this :(

**Comment by FibreFoX** [comment]
Maybe you misunderstoof me :) try to remove all settings inside your pom.xml regarding to signing. Or don't you have any options set inside <bundleArguments>-block? If so, comment them out, it is possible to create the mac.app without signing. And then you can try to `codesign` that app-bundle via CLI yourself ... maybe the codesign tool prints out some more error-codes.

**Comment by yuktikapahwa** [comment]
Can you tell me the exact position where the settings are done for codesigning the app. 
Because as far as I rememebr , I have not done any external settings in my pom to explicitly codesign a file .  It is all external to the javafx maven plugin . tHE moment the plugin starts building my application , it looks for developer id certificates and the private keys in the keychain access of my local machine. The moemnt it finds it , it starts codesiging my application and if it doesnt find it , it skips the codesigning and builds the app and the dmg. 
I think this is how it works. If you have some other information ,please let me know.

**Comment by yuktikapahwa** [comment]
For your reference , please check my maven plugin information added in my pom
`  
   <plugin>
    <groupId>com.zenjava</groupId>
    <artifactId>javafx-maven-plugin</artifactId>
    <version>8.8.0-SNAPSHOT</version>
    <configuration>
     <mainClass>com.a.application.A</mainClass>
     <deployDir>${project.basedir}/deploy</deployDir>
     <needShortcut>true</needShortcut>
     <appName>${target.name}</appName>
     <jfxMainAppJarName>a-jfx.jar</jfxMainAppJarName>

     
     <verbose>true</verbose>
     <vendor>${target.publisher}</vendor>

     
     <jfxAppOutputDir>${project.build.directory}/jfx/app</jfxAppOutputDir>

     
     <additionalAppResources>${project.build.directory}/resources/macosx/</additionalAppResources>
     <additionalBundlerResources>${project.basedir}/deploy/bundlers</additionalBundlerResources>
     <jvmArgs>
      <argument>-Xmx512m</argument>
     </jvmArgs>

    </configuration>

    
    <executions>

     

     <execution>
      <id>create-jfxjar</id>
      <phase>package</phase>
      <goals>
       <goal>build-jar</goal>
      </goals>
     </execution>

     
     <execution>
      <id>create-native</id>
      <phase>package</phase>
      <goals>
       <goal>build-native</goal>
      </goals>
     </execution>
    </executions>
   </plugin>
   <plugin>`

**Comment by FibreFoX** [comment]
Oh, I suspected some setup there, my fault of wrong thinking ;)

This is no magic of the javafx-maven-plugin, but from the javapackager itself. It tries to "find" some cerfificate matching via `security fund-certificate -c Developer ID Application: `, maybe it finds the wrong ones.

Please try to add `<mac.signing-key-developer-id-app>YOUR-ID-HERE</mac.signing-key-developer-id-app>` to the `<bundleArguments>`-block to fixate it.
If you want to play with it a bit, you may want to fill `<mac.signing-key-user-name>` and `<mac.signing-keychain>` too, for detailed information, please take a look at the Bundler-internals: https://github.com/teamfx/openjfx-8u-dev-rt/blob/83f339527558ad8a9a5180d37ecf73e1d601e42c/modules/fxpackager/src/main/java/com/oracle/tools/packager/mac/MacBaseInstallerBundler.java#L295

**Comment by yuktikapahwa** [comment]
I have got some time assigned for my self to resolve this task. Will continue on this tomorrow and try what you told me to try. Will get back to you in 15 houtrs :)

**Comment by yuktikapahwa** [comment]
Can you see in the above logs , it says that it is already signed and then it breaks after that.The most weird part is , when it is trying to code sign the .appex file , it prints that it is already code signed whereas it is not and then it breaks after that.

**Comment by FibreFoX** [comment]
Hm... so you did made that "A.app" with the bundler and (after it failing) you were running `codesign` with the same parameters, correct?

**Comment by yuktikapahwa** [comment]
yes...from the CLI , it worked . The same codesign command to codesign the appex file works from the CLI but not from the plugin.

**Comment by FibreFoX** [comment]
That's something I can work with :) and will report back when I found the problem. But one last thing: can you try to skip signing, by setting `<signBundle>` to `false` inside the `<bundleArguments>`-block? Does it work/bundle without failing? (This might be some last step before I found the error)

**Comment by yuktikapahwa** [comment]
I am sorry , where is this `<bundleArguments>` tab ? Is it inside the pom.xml ? I dont have anything like that in my pom. 
Could you please let me know the exact syntax and the position where i should palce it in my pom.So that I can go ahead and try it for you.

**Comment by FibreFoX** [comment]
```xml
<plugin>
    <groupId>com.zenjava</groupId>
    <artifactId>javafx-maven-plugin</artifactId>
    <version>8.8.0-SNAPSHOT</version>
    <configuration>
        <mainClass>com.a.application.A</mainClass>
        <deployDir>${project.basedir}/deploy</deployDir>
        <!-- ... -->
        <!-- other stuff -->
        <!-- ... -->

        <bundleArguments>
            <signBundle>false</signBundle>
        </bundleArguments>
    </configuration>
</plugin>
```

**Comment by yuktikapahwa** [comment]
Correct.
Now , as expected it built the A.app ,A.dmg and A.pkg inside the target/jfx/native folder and didnt fail . 
But there is not _CodeSignature file present inside my app becuase we explicitly ignored the code signing . Last time it was not created because the code sign was failing but this time the code sign is not even there. It also printed the following in the logs 

`[INFO] Skipping 'Mac App Store Ready Bundler' because of configuration error 'Mac App Store apps must be signed, and signing has been disabled by bundler configuration.'
Advice to fix: Either unset 'signBundle' or set 'signBundle' to true.
[INFO] Skipping 'WebStart JNLP Bundler' because of configuration error 'No OutFile Specificed'
Advice to fix: Please specify the name of the JNLP Outut file in 'jnlp.outfile'`

**Comment by FibreFoX** [comment]
This sounds wonderful, even when we are not finished yet! Just to confirm: after bundling the non-signed app, did you try to manually sign that app-bundle using the CLI-commands? And if so, did it work? (Just making sure I get it right, so sorry if I'm asking some questions again and again)

If so, this means I "only" have to track down the problem with signing, and it means that everything else now is working.

**Comment by yuktikapahwa** [comment]
Hi @FibreFoX,
I have come up with two solutions now 
1) This problem would be solved completely if you can somehow try to ignore signing the .appex insilde the PlugIns folder . That way , the bundler will sign everything and ignore this file and perfectly build the app , dmg and pkg. Then we can get rid of this problem . I am saying thAT because there is a way with which I can sign that appex file before hand .
2) If not , then I would like to let you know that I tried what you told me to try. Meaning ,I tried to codesign the app manually and then launch it . It was giving me issues launching that appex inside the app because it was treated as a misconfigured plugin because it was not sandboxed . i dont know if you understand the meaning of sanxboxed or not .Any application which is built on mac or to be run on MAC needs to be sandboxed for security purposes .
The command which you were using to sign the appex bundle is as below
`codesign -s “Developer ID Application: A LTD. (YYYYYYYYY)” --prefix com.a.application. -vvvv  <path to source>`

Whereas in order to perfectly codesign the appex bundle the command should be as below
`codesign --entitlements <Path to entitlemnts file> -s “Developer ID Application: A LTD. (YYYYYYYYY)” --prefix com.a.application. <Path to source>`

**Comment by FibreFoX** [comment]
The thing is, I am not responsible for these `codesign`-statements, thats what happens inside MacBaseInstallerBundler, which contains a method called signAppBundle, but the one with that `--entitlements`-part only gets called with the MacAppStoreBundler.

You are right, I dont know that mac-contept of "sandboxing", all I understood is, that it requires these application-bundles to be signed in a special way, so that gatekeeper (or whatever security-system is in place) gives that application access to the system.

Just a guess: if you do not add that plugin-file, creating that bundle without signing, does it work if you manually sign that thing afterwards?

**Comment by yuktikapahwa** [comment]
What I am going to try for you is  - I have removed the plugins folder from the deploy/bundlers/mac.app .Now I am letting the app build with the code signing set to true . Let us now see if the code signing succeeds and the app builds perfectly . I will not do any manual business here.

**Comment by FibreFoX** [comment]
No need to get picky, I'm not using Mac nor am I planing to do on a daily basis. All this here is done in my spare-time. I understand that it might sound like I'm letting you do the hard work, but trust me, I easily could just say "then dont sign that app". The flaw/bug is inside the JDK, I'm doing magic and monkeyworking here to get this thing to work.

If you can share a way to me, how I can create a signing certificate without the need to pay that developer account fee, I am willing to take more work on my shoulders.

I never did mac-development, nor am I planing to do so ... so please don't make our frustration about this issue (which has a comment-count over 100 ... and a high life-count) increasing more.

(sorry, but I felt to release some steam about it .... I'm still willing to nail that bug down!!!)

**Comment by yuktikapahwa** [comment]
Oh my god , I am taken aback now.
As i told you I removed the PlugIns folder and now tried to build the app just to see if everything except that is fine and I found out that the app which is built, when I double click it and try to launch it , it doesnt launch at all .
When I try to see the system logs just to find out why it is not launching , it prints the following in the system logs of MAC -

`2017-01-19 3:42:51.782 PM A.app[14678]: Failed to find library.:
2017-01-19 3:42:51.782 PM A.app[14678]: A:Failed to locate JNI_CreateJavaVM
2017-01-19 3:42:51.782 PM A.app[14678]: A:Failed to launch JVM
2017-01-19 3:42:51.784 PM com.apple.xpc.launchd[1]: (com.a.app.48032[14678]) Service exited with abnormal code: 1`
 
This means there is some other issue as well.

**Comment by yuktikapahwa** [comment]
When  I try to launch the jar present inside my application a.app from the CLI using java -jar command , it launches perfectly fine . But double clicking and running the app is not doing anything.

**Comment by FibreFoX** [comment]
hm ... the log you provided looks like there is a problem with the JRE ?! I will compare some stuff inside the MacApp-bundler and the MacAppStore-bundler, I'm suspecting some clue in there. The "problem" is the requirement on mac, not the java-part so it is expected to work when you `java -jar` your app inside :)

**Comment by yuktikapahwa** [comment]
hmm..right. 
But did u understand why this happened? I am not quite able to understand why it is not launching the app on double click :(
Please share if you got any idea or if you want me to test some other thing out .

**Comment by FibreFoX** [comment]
I now have created some apple account and registered to be part of the developer-programm. After downloading XCode, I started XCode, went to Settings > Accounts, added my fresh Apple ID and created some "Mac Development"-signing identity ... just to have some signing-key inside my local keystore. Then I added this to my pom.xml
```xml
<bundleArguments>
    <mac.signing-key-developer-id-app>Mac Development:</mac.signing-key-developer-id-app>
</bundleArguments>
```
to make sure it gets picked up for codesigning.

Everything went okay while signing, but the application crashed on execution ... I really don't know why, but calling `java -jar ` exploded for me. Don't know why it exploded, maybe because I'm running some emulated stuff.

The thing is, signing seems to work for me, but I haven't yet added some appex-file (which I even don't know how to create). I checked that I used `codesign` correctly by creating a small CLI-app inside XCode.

And I'm running out of ideas currently ... do you have any?

**Comment by JustGregory** [comment]
@yuktikapahwa 
I don't do any Mac development myself; but I wonder if the Mac system requires some specific Java JRE or JDK settings/properties/whatever. Which Java are you using? Can you check your Java classpath? Sorry if any of this has been asked before, but I know from personal experience, it's sometimes easy to overlook the simple(st) solutions...

**Comment by yuktikapahwa** [comment]
Hello 
All I could do and try is one comparison 
1) While I am not using the snapshot version of the maven plugin and setting the signing properties to false , I am able to get the app and dmg built successfully and even launch it on double click . There is no dmg issue in that . Please find the configurations of the pom below 

` 
   <plugin>
    <groupId>com.zenjava</groupId>
    <artifactId>javafx-maven-plugin</artifactId>
    <version>8.5.0</version>
    <configuration>
     <mainClass>com.a.application.A</mainClass>
     <deployDir>${project.basedir}/deploy</deployDir>
     <needShortcut>true</needShortcut>
     <appName>${target.name}</appName>
     <jfxMainAppJarName>a-jfx.jar</jfxMainAppJarName>

     
     <verbose>true</verbose>
     <vendor>${target.publisher}</vendor>

     
     <jfxAppOutputDir>${project.build.directory}/jfx/app</jfxAppOutputDir>

     
     <additionalAppResources>${project.build.directory}/resources/macosx/</additionalAppResources>

     <jvmArgs>
      <argument>-Xmx512m</argument>
     </jvmArgs>

    </configuration>

    
    <executions>

     

     <execution>
      <id>create-jfxjar</id>
      <phase>package</phase>
      <goals>
       <goal>build-jar</goal>
      </goals>
     </execution>

     
     <execution>
      <id>create-native</id>
      <phase>package</phase>
      <goals>
       <goal>build-native</goal>
      </goals>
     </execution>
    </executions>
   </plugin>`


2) Whereas when I use the new SNAPSHOT VERSION of the maven plugin with the deploy/bundlers/ support aqdded in it for Mac , I get the app built but that app is not launching and is printing "Failed to locate JNI_CreateJavaVM" in the system logs.
Please find the configurations of the pom below 
`<pluginRepositories>
    	<pluginRepository>
        	<id>oss-sonatype-snapshots</id>
        	<url>https://oss.sonatype.org/content/groups/public/</url>
        	<snapshots>
            	<enabled>true</enabled>
        	</snapshots>
    	</pluginRepository>
	</pluginRepositories>

   <plugin>
    <groupId>com.zenjava</groupId>
    <artifactId>javafx-maven-plugin</artifactId>
    <version>8.8.0-SNAPSHOT</version>
    <configuration>
     <mainClass>com.a.application.A</mainClass>
     <deployDir>${project.basedir}/deploy</deployDir>
     <needShortcut>true</needShortcut>
     <appName>${target.name}</appName>
     <jfxMainAppJarName>a-jfx.jar</jfxMainAppJarName>

     
     <verbose>true</verbose>
     <vendor>${target.publisher}</vendor>

     
     <jfxAppOutputDir>${project.build.directory}/jfx/app</jfxAppOutputDir>

     
     <additionalAppResources>${project.build.directory}/resources/macosx/</additionalAppResources>
     <additionalBundlerResources>${project.basedir}/deploy/bundlers</additionalBundlerResources>
      <bundleArguments>
            <signBundle>false</signBundle>
        </bundleArguments>
     <jvmArgs>
      <argument>-Xmx512m</argument>
     </jvmArgs>

    </configuration>

    
    <executions>

     

     <execution>
      <id>create-jfxjar</id>
      <phase>package</phase>
      <goals>
       <goal>build-jar</goal>
      </goals>
     </execution>

     
     <execution>
      <id>create-native</id>
      <phase>package</phase>
      <goals>
       <goal>build-native</goal>
      </goals>
     </execution>
    </executions>
   </plugin>
   <plugin>
`

**Comment by FibreFoX** [comment]
I have checked the difference, but their are not even near the JRE-creation: https://github.com/javafx-maven-plugin/javafx-maven-plugin/compare/javafx-maven-plugin-8.5.0...master
But you can try that by using that older plugin ... which bugs me is, that it does not run on my VM with mac ... even without signing I can't get it working, but maybe my test-project is a bit flawed ... I will try to get some executable application first, then adding some files. Will report back after I found something.

**Comment by yuktikapahwa** [comment]
Sure .
Not a problem .
If nothing turns out then we wont be able to release our app to the app store with java fx maven plugin . Thats alright. Will look for some other alternative after that. 
Lets first get sure if something can turn out good or not :)

**Comment by FibreFoX** [comment]
@yuktikapahwa just a quick question: I would like to CLOSE this issue, because it is now possible to place .appex-files into the app-bundle, remaining the problems with signing. But this might be a different problem, do you feel okay about me closing this issue and opening a follow-issue?

**Comment by yuktikapahwa** [comment]
Hi @FibreFoX , Sorry for getting back to you late.
I don't mind you closing this issue , but my only concern is that if one issue leads to another issue then there is a problem with the fix of the first issue because if the fix is correct another issue should not even be there. 
So with this snapshot build of the plugin , although we get to see the internal plugins and folders at the respective positions but still the app which is built , crashes of no reason . I am not even able to launch the application .
So , I am not sure what we should do ? Whether we should close this issue and open another one or keep this opened. Whatever is your call . 
Please note - If we mark this issue resolved , I still won't be able to use this version of the plugin because it crashes my app.

**Comment by FibreFoX** [comment]
@yuktikapahwa the changes got released with 8.8.0, will investigate about the remaining problem with signing in a different issue-ticket,
